templates/types/streaming/nextjs/app/components/ui/chat/index.ts

```diff
@@ -83,9 +83,41 @@ export type MessageAnnotation = {
   data: AnnotationData;
 };
 
+const NODE_SCORE_THRESHOLD = 0.25;
+
 export function getAnnotationData<T extends AnnotationData>(
   annotations: MessageAnnotation[],
   type: MessageAnnotationType,
 ): T[] {
-  return annotations.filter((a) => a.type === type).map((a) => a.data as T);
+  const data = annotations
+    .filter((a) => a.type === type)
+    .map((a) => a.data as T);
+
+  if (type === MessageAnnotationType.SOURCES && data.length > 0) {
+    // Filter and update the index of the source node
+    const sourceData = data[0] as SourceData;
+    if (sourceData.nodes) {
+      sourceData.nodes = preprocessSourceNodes(sourceData.nodes);
+    }
+  }
```

Comment:
better move to `index.tsx` at:

```
const sourceData = getAnnotationData<SourceData>(
    annotations,
    MessageAnnotationType.SOURCES,
  );
```

to avoid re-render might be useful to use `useMemo`

Reply:
or even better:

create a dedicated `getSourceAnnotationData` function here, use it in index.tsx with useMemo (if necessary)